src/main/java/com/learning/mfscreener/service/HistoricalNavService.java

```diff
@@ -53,104 +53,161 @@ public HistoricalNavService(
-        SchemeNameAndISIN firstByAmfi = null;
-        if (bySchemeId.isEmpty()) {
-            // discontinued Scheme
-            firstByAmfi = userSchemeDetailsService
-                    .findFirstBySchemeCode(schemeCode)
-                    .orElseThrow(
-                            () -> new SchemeNotFoundException("Fund with schemeCode " + schemeCode + " Not Found"));
-            payOut = firstByAmfi.getIsin();
-            persistSchemeInfo = true;
+        URI historicalNavUri = buildHistoricalNavUri(navDate);
+        Optional<MFSchemeEntity> bySchemeCode = this.schemeService.findBySchemeCode(schemeCode);
+        if (bySchemeCode.isPresent()) {
+            return fetchAndProcessNavData(
+                    historicalNavUri, bySchemeCode.get().getPayOut(), false, null, schemeCode, navDate);
         } else {
-            payOut = bySchemeId.get().getPayOut();
+            return handleDiscontinuedScheme(schemeCode, historicalNavUri, navDate);
         }
+    }
 
-        URI uri = UriComponentsBuilder.fromHttpUrl(historicalUrl).build().toUri();
-        String oldSchemeId = null;
+    String handleDiscontinuedScheme(Long schemeCode, URI historicalNavUri, LocalDate navDate) {
+        // discontinued scheme Isin
+        SchemeNameAndISIN schemeNameAndISIN = fetchSchemeDetails(schemeCode);
+        String payOut = schemeNameAndISIN.getIsin();
+        return fetchAndProcessNavData(historicalNavUri, payOut, true, schemeNameAndISIN, schemeCode, navDate);
+    }
+
+    String fetchAndProcessNavData(
+            URI historicalNavUri,
+            String payOut,
+            boolean persistSchemeInfo,
+            SchemeNameAndISIN schemeNameAndISIN,
+            Long schemeCode,
+            LocalDate navDate) {
         try {
-            String allNAVs = restClient.get().uri(uri).retrieve().body(String.class);
-            Reader inputString = new StringReader(Objects.requireNonNull(allNAVs));
-            try (BufferedReader br = new BufferedReader(inputString)) {
-                String lineValue = br.readLine();
-                for (int i = 0; i < 2; ++i) {
+            String allNAVsByDate = fetchHistoricalNavByCallingUri(historicalNavUri);
+            Reader inputString = new StringReader(Objects.requireNonNull(allNAVsByDate));
+            return parseNavData(inputString, payOut, persistSchemeInfo, schemeNameAndISIN, schemeCode, navDate);
+        } catch (ResourceAccessException exception) {
+            // eating as we can't do much, and it should be set when available
+            LOGGER.error("Unable to load Historical Data, downstream service is down ", exception);
+            return null;
+        }
+    }
+
+    String parseNavData(
+            Reader inputString,
+            String payOut,
+            boolean persistSchemeInfo,
+            SchemeNameAndISIN schemeNameAndISIN,
+            Long schemeCode,
+            LocalDate navDate) {
+        String oldSchemeId = null;
```

Comment:
@coderabbitai, please suggest the code to refactor parseNavData method to simplify complex logic

Reply:
@coderabbitai, the above code has ignored flag processRowByForce , please fix it

Reply:
@coderabbitai, in above code it is still processing even though we found required schemeId, thus leading to performance issues, Can you please fix it.

Reply:
@coderabbitai, I dont see processRowByForce being flipped to true, please handle it based on the condition

Reply:
@coderabbitai, the logic for flipping processRowByForce happens when processing line is a amcRow and read nextline till nonempty row occurs then if tokenize length is not 1 then flip flag as true. Please incorporate this and discard last review which fails logic

Reply:
@coderabbitai, I have handled code without loosing functionality as below, any suggestions to improve below code

```java
    private String processLines(
            BufferedReader br,
            String payOut,
            boolean persistSchemeInfo,
            SchemeNameAndISIN schemeNameAndISIN,
            Long schemeCode,
            LocalDate navDate)
            throws IOException {
        String lineValue = skipInitialLines(br);
        String oldSchemeId = null;
        String amc = null;
        String schemeType = null;
        boolean processRowByForce = false;

        while (lineValue != null && oldSchemeId == null) { // Stop processing if oldSchemeId is found
            String[] tokenize = lineValue.split(AppConstants.NAV_SEPARATOR);
            if (isAmcOrSchemeNameRow(tokenize)) {
                String tempVal = lineValue;
                lineValue = br.readLine();
                if (!StringUtils.hasText(lineValue)) {
                    lineValue = br.readLine();
                    tokenize = lineValue.split(AppConstants.NAV_SEPARATOR);
                    if (isAmcOrSchemeNameRow(tokenize)) {
                        schemeType = tempVal;
                        amc = lineValue;
                    } else {
                        amc = tempVal;
                        processRowByForce = true;
                    }
                }
                if (processRowByForce) {
                    oldSchemeId = processTokenLine(tokenize, payOut, persistSchemeInfo, amc, schemeType, oldSchemeId);
                }
            } else {
                oldSchemeId = processTokenLine(tokenize, payOut, persistSchemeInfo, amc, schemeType, oldSchemeId);
            }
            lineValue = br.readLine();
            if (!StringUtils.hasText(lineValue)) {
                lineValue = br.readLine();
            }
        }
        return finalizeSchemeInfo(oldSchemeId, schemeNameAndISIN, schemeCode, payOut);
    }
```